File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/soy/CustomSoyFileSetProvider.kt
package io.micronaut.docs.soy

import com.google.template.soy.SoyFileSet
import io.micronaut.context.annotation.Requires
import io.micronaut.core.io.ResourceLoader
import io.micronaut.views.ViewsConfiguration
import io.micronaut.views.soy.SoyFileSetProvider
import org.slf4j.LoggerFactory
import java.io.File
import java.net.URISyntaxException
import java.net.URL
import javax.inject.Singleton

@Requires(property = "spec.name", value = "soy")
//tag::clazz[]
@Singleton
class CustomSoyFileSetProvider(val resourceLoader: ResourceLoader, viewsConfiguration: ViewsConfiguration) :
    SoyFileSetProvider {
    private val folder: String
    override fun provideSoyFileSet(): SoyFileSet {
        val builder = SoyFileSet.builder()
        for (template in VIEWS) {
            resourceLoader.getResource("$folder/$template").ifPresent { url: URL ->
                try {
                    builder.add(File(url.toURI()))
                } catch (e: URISyntaxException) {
                    if (LOG.isWarnEnabled) {
                        LOG.warn(
                            "URISyntaxException raised while generating the SoyFileSet for folder {}",
                            folder, e
                        )
                    }
                }
            }
        }
        return builder.build()
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(CustomSoyFileSetProvider::class.java)
        private val VIEWS = arrayOf(
            "home.soy"
        )
    }

    init {
        folder = viewsConfiguration.folder
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/views/docs/soy/CustomSoyFileSetProvider.py
import logging

from com.google.template.soy import SoyFileSet
from jakarta.inject import Singleton
from java.io import File
from micronaut.context.annotation import Requires
from micronaut.core.io import ResourceLoader
from micronaut.views import ViewsConfiguration
from micronaut.views.soy import SoyFileSetProvider

LOG = logging.getLogger(__name__)


@Requires(property="spec.name", value="soy")
# tag::clazz[]
@Singleton
class CustomSoyFileSetProvider(SoyFileSetProvider):
    VIEWS = ["home.soy"]

    def __init__(self, views_configuration: ViewsConfiguration, resource_loader: ResourceLoader):
        self.folder = views_configuration.getFolder()
        self.resource_loader = resource_loader

    def provideSoyFileSet(self) -> SoyFileSet:
        builder = SoyFileSet.builder()
        for template in self.VIEWS:
            url = self.resource_loader.getResource(self.folder + "/" + template)
            if url.isPresent():
                try:
                    builder.add(File(url.get().toURI()))
                except Exception:
                    LOG.warning("Exception raised while generating the SoyFileSet for folder %s", self.folder, exc_info=True)
        return builder.build()
# end::clazz[]
